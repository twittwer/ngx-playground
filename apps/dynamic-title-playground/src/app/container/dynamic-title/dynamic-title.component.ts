import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { Store } from '@ngrx/store';
import { map } from 'rxjs/operators';
import { DynamicTitlePlaygroundActions } from '../../actions';

@Component({
  selector: 'dt-pg-dynamic-title',
  template: `
    <h2>
      Dynamic Title
    </h2>
    <dt-pg-title-nav *ngIf="!(isSubTitle$ | async)"></dt-pg-title-nav>
    <button (click)="setName('Foo')">Foo</button>
    <button (click)="setName('Bar')">Bar</button>
    <button (click)="setName()">Reset</button>
    <router-outlet></router-outlet>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class DynamicTitleComponent {
  public isSubTitle$ = this.activatedRoute.data.pipe(
    map((data) => data.isSubTitle),
  );

  constructor(
    private readonly activatedRoute: ActivatedRoute,
    private readonly store: Store,
  ) {}

  public setName(name?: string): void {
    this.store.dispatch(DynamicTitlePlaygroundActions.setName({ name }));
  }
}
