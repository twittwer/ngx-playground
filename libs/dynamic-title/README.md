# dynamic-title

> Define static/dynamic page titles via route configuration.

## Configuration

```typescript
@NgModule({
  imports: [
    DynamicTitleModule.forRoot({
      baseTitle: 'Application Name',
    }),
  ],
  /* ... */
})
export class AppModule {}
```

| Option         | Type                                  | Default                           | Description                                                    |
| -------------- | ------------------------------------- | --------------------------------- | -------------------------------------------------------------- |
| baseTitle\*    | `string`                              |                                   | Application name as default title or postfix.                  |
| titleFormatter | `(currentTitle, baseTitle) => string` | `${currentTitle} \| ${baseTitle}` | Formatter to build actual title based on current & base title. |

## Usage

```typescript
export const APP_ROUTES: Route[] = [
  {
    path: 'static',
    component: StaticComponent,
    data: {
      // Define static title via simple string
      title: 'Static Title',
    },
  },
  {
    path: 'dynamic',
    component: DynamicComponent,
    data: {
      // Define dynamic title via NgRx selector
      title$: selectDynamicTitle,
    },
  },
];
```
