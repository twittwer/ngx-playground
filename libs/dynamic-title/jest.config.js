module.exports = {
  name: 'dynamic-title',
  preset: '../../jest.config.js',
  coverageDirectory: '../../coverage/libs/dynamic-title',
  snapshotSerializers: [
    'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer.js',
    'jest-preset-angular/build/AngularSnapshotSerializer.js',
    'jest-preset-angular/build/HTMLCommentSerializer.js',
  ],
};
