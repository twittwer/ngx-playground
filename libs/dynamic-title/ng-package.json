{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/dynamic-title",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
