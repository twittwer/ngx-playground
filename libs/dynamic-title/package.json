{
  "name": "@twittwer/dynamic-title",
  "version": "0.0.1",
  "peerDependencies": {
    "@angular/core": ">= 8.0.0",
    "@angular/router": ">= 8.0.0",
    "@angular/platform-browser": ">= 8.0.0",
    "@ngrx/store": ">= 9.0.0",
    "@ngrx/effects": ">= 9.0.0",
    "rxjs": ">= 6.0.0"
  }
}
