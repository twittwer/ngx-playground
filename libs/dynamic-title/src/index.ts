export { DynamicTitleModule } from './lib/dynamic-title.module';

export { MinimalDynamicTitleConfig as DynamicTitleConfig } from './lib/dynamic-title-config';
export { TitleRouteData } from './lib/models/title-route-data.interface';
