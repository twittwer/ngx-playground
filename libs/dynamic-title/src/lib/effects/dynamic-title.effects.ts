import { Inject, Injectable } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { createEffect } from '@ngrx/effects';
import { select, Store } from '@ngrx/store';
import { combineLatest, Observable, of } from 'rxjs';
import { filter, map, startWith, switchMap, tap } from 'rxjs/operators';
import {
  DYNAMIC_TITLE_CONFIG,
  DynamicTitleConfig,
} from '../dynamic-title-config';
import { TitleRouteData } from '../models/title-route-data.interface';

@Injectable({ providedIn: 'root' })
export class DynamicTitleEffects {
  public updateTitle$ = createEffect(
    () =>
      this.router.events.pipe(
        filter((event) => event instanceof NavigationEnd),
        map(() => this.getRouteDataStreams()),
        switchMap((dataStreams) => combineLatest(dataStreams)),
        map((dataObjects) =>
          dataObjects.reduce(
            (prev, { title, title$ }) =>
              title$ ? { title$ } : title ? { title } : prev,
            {},
          ),
        ),
        switchMap(({ title, title$ }) => {
          if (title$) {
            return this.store.pipe(
              select(title$),
              map((selectedTitle) => selectedTitle ?? title),
              startWith(undefined),
            );
          }
          return of(title);
        }),
        map((title) => {
          const { baseTitle, titleFormatter } = this.config;
          return title ? titleFormatter(title, baseTitle) : baseTitle;
        }),
        tap((title) => {
          this.title.setTitle(title);
        }),
      ),
    {
      dispatch: false,
    },
  );

  constructor(
    private readonly router: Router,
    private readonly activatedRoute: ActivatedRoute,
    private readonly store: Store,
    private readonly title: Title,
    @Inject(DYNAMIC_TITLE_CONFIG) private readonly config: DynamicTitleConfig,
  ) {}

  private getRouteDataStreams(): Array<Observable<TitleRouteData>> {
    const dataStreams: Array<Observable<TitleRouteData>> = [];

    let route = this.activatedRoute;
    dataStreams.push(route.data);

    while (route.firstChild) {
      route = route.firstChild;
      dataStreams.push(route.data);
    }

    return dataStreams;
  }
}
