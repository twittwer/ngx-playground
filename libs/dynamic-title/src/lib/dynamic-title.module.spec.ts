import { TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import {
  defaultDynamicTitleConfig,
  DYNAMIC_TITLE_CONFIG,
  DynamicTitleConfig,
} from './dynamic-title-config';
import { DynamicTitleModule } from './dynamic-title.module';

describe('DynamicTitleModule', () => {
  it('should merge provided config with the default config', () => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        StoreModule.forRoot({}),
        EffectsModule.forRoot([]),
        DynamicTitleModule.forRoot({
          baseTitle: 'Spec',
        }),
      ],
    });

    const dynamicTitleConfig = TestBed.inject(DYNAMIC_TITLE_CONFIG);
    expect(dynamicTitleConfig).toMatchSnapshot();
    expect(dynamicTitleConfig.baseTitle).toBe('Spec');
    expect(dynamicTitleConfig.titleFormatter).toBe(
      defaultDynamicTitleConfig.titleFormatter,
    );
  });

  it('should override default with provided config', () => {
    const titleFormatter: DynamicTitleConfig['titleFormatter'] = (
      currentTitle,
      baseTitle,
    ) => `${currentTitle} - ${baseTitle}`;

    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        StoreModule.forRoot({}),
        EffectsModule.forRoot([]),
        DynamicTitleModule.forRoot({
          baseTitle: 'Spec',
          titleFormatter,
        }),
      ],
    });

    const dynamicTitleConfig = TestBed.inject(DYNAMIC_TITLE_CONFIG);
    expect(dynamicTitleConfig).toMatchSnapshot();
    expect(dynamicTitleConfig.baseTitle).toBe('Spec');
    expect(dynamicTitleConfig.titleFormatter).toBe(titleFormatter);
  });
});
