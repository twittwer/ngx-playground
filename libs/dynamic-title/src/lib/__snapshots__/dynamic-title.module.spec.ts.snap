// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicTitleModule should merge provided config with the default config 1`] = `
Object {
  "baseTitle": "Spec",
  "titleFormatter": [Function],
}
`;

exports[`DynamicTitleModule should override default with provided config 1`] = `
Object {
  "baseTitle": "Spec",
  "titleFormatter": [Function],
}
`;
